### PR TITLE
feat(rome_js_analyze): better JS suppressions

### PR DESCRIPTION
**Pull request by ematipico**
## Summary

Closes https://github.com/rome/tools/issues/3445

This PR refines the heuristic of how a suppression comment should be applied inside a mutation action. I tried to comment on the heuristic.

I also took the opportunity to move the `chain_pieces` from the formatter to the `rome_rowan` crate, and added some doc tests (which is quite lengthy, but it should be enough I hope). 

## Test Plan

I updated the test specs, to consider a new folder called `suppression`, which works the same as the current tests, but they extract only the suppression actions. 

I added some test cases, especially inside JSX, because I noticed they are the ones that caused more issues.

Plus, I run locally the suppression actions against `specs/` folder, printed the suppression actions, and checked locally that no strange code was emitted. This check was done manually and locally, I wasn't planning to update ALL the diagnostics.

If you think there is some edge case that requires some attention, please paste some snippets and I will try to add them to the test suite.

### CODE REVIEW

File: crates/rome_js_analyze/src/suppression_action.rs [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+///
+/// Due to the nature of JSX, sometimes the current token might contain text that contains
+/// some newline. In case that happens, we choose that token.
+fn find_fist_token_with_newline(token: JsSyntaxToken) -> Option<(JsSyntaxToken, bool)> {
+    let mut current_token = token;
+    let mut is_at_root = false;
+    loop {
+        let trivia = current_token.leading_trivia();
+        if trivia.pieces().any(|trivia| trivia.is_newline())
+            || current_token.text_trimmed().contains('\n')
```
Comment (MichaReiser) [inline crates/rome_js_analyze/src/suppression_action.rs:145]: This won't work on systems with `\r` line breaks.

File: crates/rome_js_analyze/src/suppression_action.rs [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+    });
+
+    if let Some((first_token_with_newline, is_at_root)) = first_token_with_newline {
+        // we check if the token that has the newline is inside a JSX element: JsxOpeningElement or JsxSelfClosingElement
+        let current_jsx_element = first_token_with_newline.parent().and_then(|parent| {
+            if AnyJsxElement::can_cast(parent.kind()) || JsxText::can_cast(parent.kind()) {
+                Some(parent)
+            } else {
+                None
+            }
```
Comment (MichaReiser) [inline crates/rome_js_analyze/src/suppression_action.rs:44]: Is it necessary to also test for templates here as they can contain new lines too?

File: crates/rome_js_analyze/src/suppression_action.rs [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+        suppression_text,
+        diagnostic_text_range,
+    } = payload;
+    // retrieve the most suited, most left token where the diagnostics was emitted
+    let original_token = peek_token_from_offset(token_offset, diagnostic_text_range);
+
+    // considering that our suppression system works via lines, we need to look for the first newline,
+    // so we can place the comment there
+    let first_token_with_newline = original_token.as_ref().map(|original_token| {
+        match find_fist_token_with_newline(original_token.clone()) {
```
Comment (MichaReiser) [inline crates/rome_js_analyze/src/suppression_action.rs:31]: first?
```suggestion
        match find_first_token_with_newline(original_token.clone()) {
```

File: crates/rome_js_analyze/src/suppression_action.rs [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+                        &AnyJsxChild::JsxText(current_element.clone()),
+                        &[
+                            AnyJsxChild::JsxText(current_element.detach()),
+                            AnyJsxChild::JsxExpressionChild(jsx_comment),
+                        ],
+                    );
+                }
+            } else {
+                let new_token = if !is_at_root {
+                    first_token_with_newline.with_leading_trivia([
```
Comment (MichaReiser) [inline crates/rome_js_analyze/src/suppression_action.rs:91]: Does this replace any existing trivia on the token?

Comment (ematipico) [reply crates/rome_js_analyze/src/suppression_action.rs]: That's fixed and added a test case with existing comments.

File: crates/rome_js_analyze/src/suppression_action.rs (hunk 2 above)
Comment (ematipico) [reply crates/rome_js_analyze/src/suppression_action.rs:47]: I added a specific branch just for template literals, their logic is different. A test case has been added.

File: crates/rome_js_analyze/src/suppression_action.rs (hunk 1 above)
Comment (ematipico) [reply crates/rome_js_analyze/src/suppression_action.rs:153]: I added a new check

File: crates/rome_js_analyze/tests/spec_tests.rs [hunk 5]
```diff
@@ -91,7 +106,12 @@ fn write_analysis_to_snapshot(
     rome_js_analyze::analyze(FileId::zero(), &root, filter, &options, |event| {
         if let Some(mut diag) = event.diagnostic() {
             for action in event.actions() {
-                if !action.is_suppression() {
+                if check_action_type.is_suppression() {
```
[the remark this answers is not in the document]
Comment (ematipico) [reply crates/rome_js_analyze/tests/spec_tests.rs:109]: Doing so will add the suggested actions to the snapshot of the suppressions, which is not what I wanted to avoid

File: crates/rome_js_analyze/src/suppression_action.rs [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+    loop {
+        let trivia = current_token.leading_trivia();
+        if trivia.pieces().any(|trivia| trivia.is_newline())
+            || current_token.text_trimmed().contains('\n')
+        {
+            break;
+        } else if let Some(token) = current_token.prev_token() {
+            current_token = token;
+            continue;
+        } else if let Some(parent_token) = current_token
```
[the remark this answers is not in the document]
Comment (ematipico) [reply crates/rome_js_analyze/src/suppression_action.rs:172]: Very true! I removed the branch

File: crates/rome_js_analyze/tests/spec_tests.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (leops) [reply crates/rome_js_analyze/tests/spec_tests.rs:136]: Sorry I messed up factorizing the logic expression, I think it's actually `!(check_action_type.is_suppression() ^ action.is_suppression())`

File: crates/rome_rowan/src/syntax_rewriter.rs [hunk 7]
```diff
@@ -255,3 +255,145 @@ mod tests {
...
+///
+/// let leading_trivia = chain_pieces(
+///     first_token.leading_trivia().pieces(),
+///     third_token.leading_trivia().pieces()
+/// );
+///
+/// let new_first_token = first_token.with_leading_trivia_pieces(leading_trivia);
+///
+/// let new_token = format!("{:?}", new_first_token);
+/// assert_eq!(new_token, "LET_TOKEN@0..10 \"let\" [Whitespace(\"\\n\\t \"), Whitespace(\";\")] [Whitespace(\" \\t\\t\")]")
```
Comment (MichaReiser) [inline crates/rome_rowan/src/syntax_rewriter.rs:297]: Using `unwrap` in dotests is discouraged because readers may end up copying the code including the `unwrap` calls. That's why it is recommended to wrap the test in a `main` function and hide these lines by prefixing them with `#`.

```suggestion
/// # use rome_rowan::raw_language::{RawLanguageKind, RawSyntaxTreeBuilder};
/// # use rome_rowan::{chain_pieces, TriviaPiece, TriviaPieceKind, SyntaxResult};
/// 
/// # fn main() -> SyntaxResult<()> {
/// let mut builder = RawSyntaxTreeBuilder::new();
///
/// let mut node = RawSyntaxTreeBuilder::wrap_with_node(RawLanguageKind::ROOT, |builder| {
///     builder.token_with_trivia(
///         RawLanguageKind::LET_TOKEN,
///         "\n\t let \t\t",
///         &[TriviaPiece::whitespace(3)],
///         &[TriviaPiece::whitespace(3)],
///     );
///     builder.token(RawLanguageKind::STRING_TOKEN, "a");
///     builder.token_with_trivia(
///         RawLanguageKind::SEMICOLON_TOKEN,
///         "; \t\t",
///         &[TriviaPiece::whitespace(1)],
///         &[TriviaPiece::whitespace(1)],
///     );
/// });
/// let mut tokens = node.tokens();
/// let first_token = tokens.next()?;
/// let second_token = tokens.next()?;
/// let third_token = tokens.next()?;
///
/// let leading_trivia = chain_pieces(
///     first_token.leading_trivia().pieces(),
///     third_token.leading_trivia().pieces()
/// );
///
/// let new_first_token = first_token.with_leading_trivia_pieces(leading_trivia);
///
/// let new_token = format!("{:?}", new_first_token);
/// assert_eq!(new_token, "LET_TOKEN@0..10 \"let\" [Whitespace(\"\\n\\t \"), Whitespace(\";\")] [Whitespace(\" \\t\\t\")]")
/// # Ok(())
/// # }
```

File: .cargo/config.toml [hunk 8]
```diff
@@ -1,5 +1,5 @@
 [alias]
-lint = "clippy --workspace --all-targets --verbose -- --deny warnings"
+lint = "clippy --workspace --all-targets --verbose -- --deny warnings -W clippy::cargo -W clippy::dbg_macro -A clippy::cargo_common_metadata"
```
Comment (MichaReiser) [inline .cargo/config.toml:2]: Nit: Unrelated to this PR

File: crates/rome_js_analyze/src/utils/batch.rs [hunk 9]
```diff
@@ -22,17 +22,17 @@ pub trait JsBatchMutation {
     fn remove_js_object_member(&mut self, parameter: &AnyJsObjectMember) -> bool;
 
     /// It attempts to add a new element after the given element
-    fn add_jsx_element_after_element(
+    fn add_jsx_elements_after_element(
         &mut self,
         after_element: &AnyJsxChild,
-        new_element: &AnyJsxChild,
+        new_element: &[AnyJsxChild],
```
Comment (MichaReiser) [inline crates/rome_js_analyze/src/utils/batch.rs:28]: Nit: `new_elements`

File: crates/rome_js_analyze/src/utils/batch.rs [hunk 10]
```diff
@@ -221,40 +221,40 @@ impl JsBatchMutation for BatchMutation<JsLanguage> {
                 for element in old_list {
                     new_items.push(element.clone());
                     if element == *after_element {
-                        new_items.push(new_element.clone());
+                        new_items.extend(new_element.to_vec());
```
Comment (MichaReiser) [inline crates/rome_js_analyze/src/utils/batch.rs:224]: `extend_from_slice(new_elements)`

OR: change `new_elements` to `impl Iterator<Item=AnyJsxChild>` and call it with `items.cloned()`

File: crates/rome_js_analyze/src/utils/batch.rs [hunk 11]
```diff
@@ -221,40 +221,40 @@ impl JsBatchMutation for BatchMutation<JsLanguage> {
...
+        new_elements: &[AnyJsxChild],
     ) -> bool {
         let old_list = after_element.syntax().parent().and_then(JsxChildList::cast);
         if let Some(old_list) = &old_list {
             let jsx_child_list = {
                 let mut new_items = vec![];
                 for element in old_list {
                     if element == *after_element {
-                        new_items.push(new_element.clone());
+                        new_items.extend(new_elements.to_vec());
```
Comment (MichaReiser) [inline crates/rome_js_analyze/src/utils/batch.rs:249]: `extend_from_slice` 

OR: change `new_elements` to `impl Iterator<Item=AnyJsxChild>` and call it with `items.cloned()`

File: crates/rome_js_analyze/src/lib.rs [hunk 12]
```diff
@@ -262,6 +258,7 @@ mod tests {
 
                 for action in signal.actions() {
                     let new_code = action.mutation.commit();
+                    dbg!(&new_code);
```
Comment (MichaReiser) [inline crates/rome_js_analyze/src/lib.rs:261]: Remove?

File: crates/rome_js_analyze/src/suppression_action.rs [hunk 13]
```diff
@@ -0,0 +1,220 @@
...
+use rome_analyze::SuppressionCommentEmitterPayload;
+use rome_js_factory::make::{jsx_expression_child, token};
+use rome_js_syntax::jsx_ext::AnyJsxElement;
+use rome_js_syntax::{
+    AnyJsxChild, JsLanguage, JsSyntaxKind, JsSyntaxToken, JsTemplateElement, JsxChildList,
+    JsxElement, JsxOpeningElement, JsxSelfClosingElement, JsxText, TextRange, T,
+};
+use rome_rowan::{AstNode, AstNodeExt, TokenAtOffset, TriviaPieceKind};
+
+/// We now try to "guess" the token where to apply the suppression comment.
```
Comment (MichaReiser) [inline crates/rome_js_analyze/src/suppression_action.rs:11]: Can we add some documentation what the function does (it does more than just guessing?) 

```
Adds a suppression comment.

The implementation guesses the token to which it applies the suppression comment. 
```

Using `we` in documentatation can be confusing. It's not me and you doing the guessing but the implementation/algorithm.

File: crates/rome_js_analyze/src/suppression_action.rs [hunk 14]
```diff
@@ -0,0 +1,220 @@
...
+/// function f() {
+///     return <div
+///     ><img /> {/* <--- diagnostic emitted in this line */}
+///     </div>
+/// }
+/// ```
+///
+/// In these case it's best to peek the right token, because it belongs to the node where error actually occurred,
+/// and becomes easier to add the suppression comment.
+fn peek_token_from_offset(
```
Comment (MichaReiser) [inline crates/rome_js_analyze/src/suppression_action.rs:203]: What's the meaning of `peek` in this function. `Peek` is commonly used with iterators to look ahead without (seemingly) progressing the iterator. 

This isn't the case for this function (it even consumes `TokenAtOffset`). How about naming it `get_closest_token_for_diagnostic(...)`

File: crates/rome_js_analyze/src/suppression_action.rs [hunk 15]
```diff
@@ -0,0 +1,220 @@
...
+                    (TriviaPieceKind::Newline, "\n"),
+                ])
+            };
+            mutation.replace_token_transfer_trivia(first_token_with_newline, new_token);
+        }
+    }
+}
+
+/// It checks if the current token has leading trivia newline. If not, it
+/// it peeks the previous token and recursively call itself.
```
Comment (MichaReiser) [inline crates/rome_js_analyze/src/suppression_action.rs:139]: /// Finds the first token, starting with the current token and traversing backwards, that has a leading newline trivia.

File: crates/rome_js_analyze/src/suppression_action.rs [hunk 16]
```diff
@@ -0,0 +1,220 @@
...
+/// `${` and if so, we stop there.
+fn find_first_token_with_newline(token: JsSyntaxToken) -> Option<(JsSyntaxToken, bool)> {
+    let mut current_token = token;
+    let mut should_insert_leading_newline = false;
+    loop {
+        let trivia = current_token.leading_trivia();
+        if trivia.pieces().any(|trivia| trivia.is_newline())
+            || current_token.text_trimmed().contains('\n')
+            || current_token.text_trimmed().contains("\r\n")
+        {
```
Comment (MichaReiser) [inline crates/rome_js_analyze/src/suppression_action.rs:154]: This test isn't sufficient to handle all JS new line characters:

```
* `\n`
* `\r`
* `\r\n`
```

```suggestion
current_token.text_trimmed().contains(['\n', '\r'])
```


And can you add a test for multiline strings too:

```
"a b\
  cd"
```

I recommend scoping the check for newlines inside of the text to the few token kinds that allow newlines rather than applying it everywhere. The tokens that support newlines in the middle of the text are:

* string literal
* JsxText
* TEMPLATE_CHUNK

File: crates/rome_js_analyze/src/suppression_action.rs [hunk 17]
```diff
@@ -0,0 +1,220 @@
...
+            break;
+        } else if matches!(current_token.kind(), JsSyntaxKind::DOLLAR_CURLY)
+            && current_token
+                .parent()
+                .map(|p| JsTemplateElement::can_cast(p.kind()))
```
Comment (MichaReiser) [inline crates/rome_js_analyze/src/suppression_action.rs:159]: Is it necessary to also handle `TsTemplateElement`?

File: crates/rome_js_analyze/src/suppression_action.rs [hunk 18]
```diff
@@ -0,0 +1,220 @@
...
+                .unwrap_or_default()
+        {
+            if let Some(next_token) = current_token.next_token() {
```
Comment (MichaReiser) [inline crates/rome_js_analyze/src/suppression_action.rs:162]: Why is this path taking the `next_token`. The documentation mentions that the algorithm traverses backwards. 

Is there this path traverses forward and the next iteration traverses backwards again, resulting in an infinite loop?

File: crates/rome_js_analyze/src/suppression_action.rs [hunk 19]
```diff
@@ -0,0 +1,220 @@
...
+    let mut should_insert_leading_newline = false;
+    loop {
+        let trivia = current_token.leading_trivia();
+        if trivia.pieces().any(|trivia| trivia.is_newline())
+            || current_token.text_trimmed().contains('\n')
+            || current_token.text_trimmed().contains("\r\n")
+        {
+            break;
+        } else if matches!(current_token.kind(), JsSyntaxKind::DOLLAR_CURLY)
+            && current_token
```
Comment (MichaReiser) [inline crates/rome_js_analyze/src/suppression_action.rs:157]: Nit: `current_token.parent().kind() == Some(JsSyntaxKind::JS_TEMPATE_ELEMENT)`

File: crates/rome_js_analyze/src/suppression_action.rs [hunk 20]
```diff
@@ -0,0 +1,220 @@
...
+                continue;
+            }
+        } else if let Some(token) = current_token.prev_token() {
+            current_token = token;
+            continue;
+        } else {
+            should_insert_leading_newline = true;
+            break;
+        }
+    }
```
Comment (MichaReiser) [inline crates/rome_js_analyze/src/suppression_action.rs:175]: Nit:

There are a few unnecessary `continue` calls and explicitly returning a value from the loop can improve readabilty

```suggestion
    let should_insert_leading_newline = loop {
        let trivia = current_token.leading_trivia();
        if trivia.pieces().any(|trivia| trivia.is_newline())
            || current_token.text_trimmed().contains('\n')
            || current_token.text_trimmed().contains("\r\n")
        {
            break false;
        } else if matches!(current_token.kind(), JsSyntaxKind::DOLLAR_CURLY)
            && current_token
                .parent()
                .map(|p| JsTemplateElement::can_cast(p.kind()))
                .unwrap_or_default()
        {
            if let Some(next_token) = current_token.next_token() {
                current_token = next_token;
                break false;
            } 
        } else if let Some(token) = current_token.prev_token() {
            current_token = token;
        } else {
            break true;
        }
    }

```

File: crates/rome_js_analyze/src/suppression_action.rs [hunk 21]
```diff
@@ -0,0 +1,220 @@
...
+/// some newline. In case that happens, we choose that token.
+///
+/// Due to the nature of JavaScript templates, we also check if the tokens we browse are
+/// `${` and if so, we stop there.
+fn find_first_token_with_newline(token: JsSyntaxToken) -> Option<(JsSyntaxToken, bool)> {
+    let mut current_token = token;
+    let mut should_insert_leading_newline = false;
+    loop {
+        let trivia = current_token.leading_trivia();
+        if trivia.pieces().any(|trivia| trivia.is_newline())
```
Comment (MichaReiser) [inline crates/rome_js_analyze/src/suppression_action.rs:151]: You explicitly handle multiline comments at the beginning of the file. Isn't it necessary to handle multiline comments here too, e.g. to support

```
if (/ ** some weird comment
    */ a == b) {

}
```

File: crates/rome_js_analyze/tests/suppression/correctness/noUndeclaredVariables.ts.snap [hunk 22]
```diff
@@ -0,0 +1,28 @@
...
+export type Invalid<S extends number> = `Hello ${T}`
+```
+
+# Diagnostics
+```
+noUndeclaredVariables.ts:1:50 lint/correctness/noUndeclaredVariables  FIXABLE  ━━━━━━━━━━━━━━━━━━━━━
+
+  ! The T variable is undeclared
+  
+  > 1 │ export type Invalid<S extends number> = `Hello ${T}`
```
Comment (MichaReiser) [inline crates/rome_js_analyze/tests/suppression/correctness/noUndeclaredVariables.ts.snap:16]: Can you try a multiline template and a regular JS template

```
let a = `Output of "rome rage":
  formatter enabled: ${formatter == true}
  linter: ${linter}
`;  
```

File: crates/rome_js_analyze/tests/suppression/correctness/noArrayIndexKey.jsx.snap [hunk 23]
```diff
@@ -0,0 +1,178 @@
...
+  i The order of the items may change, and this also affects performances and component state.
+  
+  i Check the React documentation. 
+  
+  i Safe fix: Suppress rule lint/correctness/noArrayIndexKey
+  
+    15 15 │   something.forEach((Element, index) => {
+    16 16 │       return <List
+    17    │ - ········><div·key={index}>foo</div>
+       17 │ + ········//·rome-ignore·lint/correctness/noArrayIndexKey:·suppressed·
```
Comment (MichaReiser) [inline crates/rome_js_analyze/tests/suppression/correctness/noArrayIndexKey.jsx.snap:170]: This is incorrect. It places the suppression comment inside of the JsxText.

File: crates/rome_js_analyze/tests/suppression/correctness/noArrayIndexKey.jsx.snap [hunk 24]
```diff
@@ -0,0 +1,178 @@
...
+       18 │ + ><div·key={index}>foo</div>
+    18 19 │       </List>;
+    19 20 │   });
```
Comment (MichaReiser) [inline crates/rome_js_analyze/tests/suppression/correctness/noArrayIndexKey.jsx.snap:173]: Can we add an example of a multiline jsx element

```
<div 
  className="SomeManyClasses"
  onClick={(event) => {
    console.log(event.current.target);
  }}
  style="color: red"
  aria-role="button"
>
Some content
</div>
```

File: crates/rome_js_analyze/src/suppression_action.rs [hunk 25]
```diff
@@ -0,0 +1,220 @@
...
+            } else {
+                new_token = new_token.with_leading_trivia([
+                    (
+                        TriviaPieceKind::SingleLineComment,
+                        format!("// {}: suppressed ", suppression_text).as_str(),
+                    ),
+                    (TriviaPieceKind::Newline, "\n"),
+                ])
+            };
+            mutation.replace_token_transfer_trivia(first_token_with_newline, new_token);
```
Comment (MichaReiser) [inline crates/rome_js_analyze/src/suppression_action.rs:133]: The `replace_token_transfer_trivia` implementation looks incorrect. Isn't it applying the trivia of the `prev_token` (leading) and `next_token` (trailing) twice 

```
        let leading_trivia: Vec<_> = prev_token
            .leading_trivia()
            .pieces()
            .chain(prev_token.leading_trivia().pieces())
            .map(|piece| TriviaPiece::new(piece.kind(), TextSize::of(piece.text())))
            .collect();

        let trailing_trivia: Vec<_> = next_token
            .trailing_trivia()
            .pieces()
            .chain(next_token.trailing_trivia().pieces())
            .map(|piece| TriviaPiece::new(piece.kind(), TextSize::of(piece.text())))
            .collect();

```

Nit: replace_token_transfer_trivia can now use `chain_trivia_pieces`

**Review by MichaReiser — CHANGES_REQUESTED** [review]

**Comment by MichaReiser** [comment]
Requesting changes because:
* It seems to me that the mutation adds the suppression comment as JSX Text for one of the test cases
* `replace_token_transfer_trivia` doesn't seem to be correctly implemented. Can you add tests for testing the path where trivia is merged (normal comments, multiline comments, etc)

File: crates/rome_js_analyze/tests/spec_tests.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (ematipico) [reply crates/rome_js_analyze/tests/spec_tests.rs:136]: Same result 😅

File: crates/rome_js_analyze/src/suppression_action.rs (hunk 18 above)
Comment (ematipico) [reply crates/rome_js_analyze/src/suppression_action.rs:162]: That's because we can't place a suppression comment **before** `${`, but we are forced to place the suppression comment **inside** template expression, and thus we need to pick the next token after `${`.

Theortelly, I think this will never go into a for an infinite `next_token()` that was actually processed in the previous iteration.

File: crates/rome_js_analyze/src/suppression_action.rs (hunk 19 above)
Comment (ematipico) [reply crates/rome_js_analyze/src/suppression_action.rs:157]: I prefer the current code, `parent` might be misleading at first glance

File: crates/rome_js_analyze/tests/suppression/correctness/noArrayIndexKey.jsx.snap (hunk 23 above)
Comment (ematipico) [reply crates/rome_js_analyze/tests/suppression/correctness/noArrayIndexKey.jsx.snap:170]: The comment is placed _inside_ `div` and before the `>` that goes to a new line. The syntax is correct (otherwise the test suite would have failed). I would argue that the placement of the comment is not ideal but:
- the source code is itself an edge case
- the placement of the comment is in line with how [suppression comments work](https://docs.rome.tools/playground/?lintRules=all&code=cwBvAG0AZQB0AGgAaQBuAGcALgBmAG8AcgBFAGEAYwBoACgAKABFAGwAZQBtAGUAbgB0ACwAIABpAG4AZABlAHgAKQAgAD0APgAgAHsACgAgACAAIAAgAHIAZQB0AHUAcgBuACAAPABMAGkAcwB0AAoAIAAgACAAIAAgACAAIAAgAC8ALwAgAHIAbwBtAGUALQBpAGcAbgBvAHIAZQAgAGwAaQBuAHQALwBjAG8AcgByAGUAYwB0AG4AZQBzAHMALwBuAG8AQQByAHIAYQB5AEkAbgBkAGUAeABLAGUAeQA6ACAAcwB1AHAAcAByAGUAcwBzAGUAZAAgAAoAPgA8AGQAaQB2ACAAawBlAHkAPQB7AGkAbgBkAGUAeAB9AD4AZgBvAG8APAAvAGQAaQB2AD4ACgAgACAAIAAgADwALwBMAGkAcwB0AD4AOwAKAH0AKQA%3D)

File: crates/rome_js_analyze/src/suppression_action.rs (hunk 25 above)
Comment (ematipico) [reply crates/rome_js_analyze/src/suppression_action.rs:133]: This is done inside `replace_token_transfer_trivia` function. I added some other test cases to show that.

File: crates/rome_rowan/src/trivia.rs [hunk 26]
```diff
@@ -0,0 +1,153 @@
+//! Module that exports APIs to help working with trivia
```
Comment (MichaReiser) [inline crates/rome_rowan/src/trivia.rs:1]: I recommend to move the implementation into the existing `syntax/trivia` file

File: crates/rome_rowan/src/trivia.rs [hunk 27]
```diff
@@ -0,0 +1,153 @@
...
+///         &[TriviaPiece::whitespace(3)],
+///     );
+///     builder.token(RawLanguageKind::STRING_TOKEN, "a");
+///     builder.token_with_trivia(
+///         RawLanguageKind::SEMICOLON_TOKEN,
+///         "; \t\t",
+///         &[TriviaPiece::whitespace(1)],
+///         &[TriviaPiece::whitespace(1)],
+///     );
+///  });
```
Comment (MichaReiser) [inline crates/rome_rowan/src/trivia.rs:33]: Can we use `SyntaxToken::new_detached` instead

File: crates/rome_js_analyze/src/suppression_action.rs (hunk 15 above)
Comment (MichaReiser) [reply crates/rome_js_analyze/src/suppression_action.rs:172]: If you don't like the suggestion, replace the `it it` with `it`

File: crates/rome_js_analyze/tests/suppression/a11y/useKeyWithClickEvents.jsx.snap [hunk 28]
```diff
@@ -0,0 +1,47 @@
...
+```js
+<div
+    className="SomeManyClasses"
+    onClick={(event) => {
+        console.log(event.current.target);
+    }}
+    style="color: red"
+>
+    Some content
+</div>
```
Comment (MichaReiser) [inline crates/rome_js_analyze/tests/suppression/a11y/useKeyWithClickEvents.jsx.snap:15]: Okay. that was a bad example on my side. I thought the rule would trigger on the `onClick` attribute and not on the whole element. 

Can you add a test for


```suggestion
<button
    className="SomeManyClasses"
    onClick={(event) => console.log(event.ctrlKey == true)}
    style="color: red"
>
    Some content
</button>
```

File: crates/rome_js_analyze/src/utils/batch.rs (hunk 9 above)
Comment (MichaReiser) [reply crates/rome_js_analyze/src/utils/batch.rs:28]:
```suggestion
        new_elements: &[AnyJsxChild],
```

File: crates/rome_js_analyze/src/suppression_action.rs [hunk 29]
```diff
@@ -0,0 +1,314 @@
...
+                    );
+                } else if let Some(current_element) = JsxText::cast_ref(&current_jsx_element) {
+                    // We want to add an additional JsxText to keep the indentation
+                    let indentation_text = make_indentation_from_jsx_element(&current_element);
+                    mutation.add_jsx_elements_after_element(
+                        &AnyJsxChild::JsxText(current_element),
+                        &[
+                            AnyJsxChild::JsxExpressionChild(jsx_comment),
+                            AnyJsxChild::JsxText(indentation_text),
+                        ],
```
Comment (MichaReiser) [inline crates/rome_js_analyze/src/suppression_action.rs:92]: I recommend changing the `new_elements` argument of `add_jsx_elements_after_element` to `IntoIterator<Item=..>` instead because:

* It removes the need to pass a slice here (you can write `[AnyJsxChild::...]` directly without the need for the `&`
* It avoids cloning the elements inside of `add_jsx_elements_after_element`


Same for `add_jsx_elements_before_element`

File: crates/rome_js_analyze/src/utils/batch.rs [hunk 30]
```diff
@@ -209,52 +215,75 @@ impl JsBatchMutation for BatchMutation<JsLanguage> {
...
+                let mut old_elements = old_list.into_iter().peekable();
+
+                while let Some(next_element) = old_elements.peek() {
+                    if next_element == before_element {
+                        break;
                     }
-                    new_items.push(element.clone());
+                    // SAFETY: the peek() makes sure the element exists
+                    let next_element = old_elements.next().unwrap();
+                    new_items.push(next_element);
```
Comment (MichaReiser) [inline crates/rome_js_analyze/src/utils/batch.rs:277]: Can't we push the peeked element instead so that we avoid the `unwrap` call 

```suggestion
                    new_items.push(next_element);
                    old_elements.next();
```

**Review by MichaReiser — APPROVED** [review]
